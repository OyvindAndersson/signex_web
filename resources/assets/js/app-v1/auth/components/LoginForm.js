import React from 'react'
import PropTypes from 'prop-types'

/**
 * Presentational part of the login page, taking care
 * of form and action
 */
export default class LoginForm extends React.Component {
    constructor(props){
        super(props)

        this.handleChange = this.handleChange.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)
        this.state = {
            email: "",
            password: ""
        }
    }
    handleChange(e){
        let name = e.target.name
        let value = e.target.value

        this.setState({
            [name]: value
        })
    }
    handleSubmit(e){
        e.preventDefault()
        
        var {email, password} = this.state
        this.props.authLoginUser({email, password})
    }
    
    render() {
        const {isVerifyingLogin} = this.props
        return(
            <div className="container">
                <div className="row justify-content-md-center">
                    <div className="col-md-6 col-lg-4">

                        <div className="card">
                         <img className="card-img-top" 
                            src="https://stormideaswus.blob.core.windows.net/headerjunction/2014/91/6d2380c6-00e3-4f58-a911-c8aa98afa460.jpg" 
                            alt="Card image cap" />
                            {isVerifyingLogin ? (
                                <div className="progress">
                                    <div 
                                        className="progress-bar progress-bar-striped progress-bar-animated bg-warning" 
                                        role="progressbar" 
                                        aria-valuenow="100" 
                                        aria-valuemin="0" 
                                        aria-valuemax="100" 
                                        style={{width: 100+"%"}}>
                                    </div>
                                </div>
                                ) : null }
                            <div className="card-body">
                                <h4 className="card-title">Login</h4>

                                <form onSubmit={this.handleSubmit}>
                                    <div className="form-group">
                                        <label htmlFor="inputEmail">Email address</label>
                                        <input type="email" 
                                            className="form-control form-control-lg" 
                                            id="inputEmail" 
                                            name="email"
                                            value={this.state.email} 
                                            onChange={this.handleChange}
                                            aria-describedby="emailHelp" 
                                            placeholder="Enter email" />
                                        <small id="emailHelp" className="form-text text-muted">
                                            We'll never share your email with anyone else.
                                        </small>
                                    </div>
                                    <div className="form-group">
                                        <label htmlFor="inputPassword">Password</label>
                                        <input type="password" 
                                            className="form-control form-control-lg" 
                                            id="inputPassword" 
                                            name="password"
                                            value={this.state.password}
                                            onChange={this.handleChange}
                                            placeholder="Password" />
                                    </div>

                                    <button type="submit" className="btn btn-primary">Login</button>
                                </form>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        )
        
    }
}
LoginForm.propTypes = {
    authLoginUser: PropTypes.func.isRequired
}
