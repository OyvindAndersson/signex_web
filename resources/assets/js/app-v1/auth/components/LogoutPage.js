import React from 'react'
import {connect} from 'react-redux'
import {Redirect} from 'react-router-dom'

import types from 'Auth/actionTypes'
import constants from 'Auth/constants'
import { isClientAuthenticated } from 'Auth/selectors'
import { logoutAction } from 'Auth/actions';

/**
 * Logout page container
 */
class LogoutPage extends React.Component {
    componentDidMount(){
        if(this.props.logout){
            localStorage.setItem('isAuthenticated', false)
            this.props.logout()
        }
    }
    componentWillReceiveProps(props){
        if(!props.isAuthenticated) {
            console.log('LogoutPage - isAuthenticated ', `[${props.isAuthenticated}]`)
        }
    }
    render(){
        const {isAuthenticated} = this.props
        if(!isAuthenticated){
            return(<Redirect to={constants.LOGIN_PATH}/>)
        }
        else{
            return(<div>Logging out...</div>)
        }
        
    }
}

function mapStateToProps(state){
    return {
        isAuthenticated: isClientAuthenticated(state)
    }
}

function mapDispatchToProps(dispatch){
    return {
        logout: () => dispatch(logoutAction())
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(LogoutPage)