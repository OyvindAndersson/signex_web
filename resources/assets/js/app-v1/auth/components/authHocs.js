import React from 'react'

const defaults = {
    LoadingComponent: () => null, // dont render anything while working
    ElseComponent: () => null, // dont render anything on failure of the predicate
    wrapperDisplayName: 'AuthWrapper'
  }

  /**
   * Only render WrappedComponent when authenticated.
   * If not authed, FailureComponent will render.
   * If login is dispatched, or verify-token is
   * dispatched, LoadingComponent is rendered
   * 
   * @param {*} args 
   */
export const onlyAuth = (args) => {
    const { LoadingComponent, ElseComponent, wrapperDisplayName } = {
        ...defaults,
        ...args
      }

    function wrapComponent(WrappedComponent){
        const displayName = WrappedComponent.displayName || WrappedComponent.name || 'Component'

        class AuthWrapper extends React.Component {
            render(){
                const {isAuthenticated, isVerifyingLogin, isVerifyingToken} = this.props

                if(isAuthenticated && !isVerifyingLogin && !isVerifyingToken){
                    return <WrappedComponent {...this.props} />
                } else if(isVerifyingLogin || isVerifyingToken){
                    return <LoadingComponent {...this.props} />
                } else {
                    return <ElseComponent {...this.props} />
                }
            }
        }

        return AuthWrapper
    }

    return wrapComponent
}

/**
 * Only render Wrapped component when not authenticated.
 * If authenticated, the ElseComponent will render.
 * 
 * @param {*} args 
 */
export const exceptAuth = (args) => {
    const { LoadingComponent, ElseComponent, wrapperDisplayName } = {
        ...defaults,
        ...args
      }

    function wrapComponent(WrappedComponent){
        const displayName = WrappedComponent.displayName || WrappedComponent.name || 'Component'

        class AuthWrapper extends React.Component {
            render(){
                const {isAuthenticated, isVerifyingLogin, isVerifyingToken} = this.props

                if(!isAuthenticated && !isVerifyingLogin && !isVerifyingToken){
                    return <WrappedComponent {...this.props} />
                } else if(isVerifyingLogin || isVerifyingToken){
                    return <LoadingComponent {...this.props} />
                } else {
                    return <ElseComponent {...this.props} />
                }
            }
        }

        return AuthWrapper
    }

    return wrapComponent
}