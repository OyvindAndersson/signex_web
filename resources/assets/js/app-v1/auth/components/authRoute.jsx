import React from 'react'
import {Route, Redirect} from 'react-router-dom'
import {connect} from 'react-redux'
import constants from '../constants'
import jwtDecode from 'jwt-decode'
import moment from 'moment'

import { authUserToken } from '../actions'
import types from '../actionTypes'

  /**
   * Auth Route Wrapper
   * 
   * Wraps a react-router Route. Yields access based upon
   * authenticated status and role, if defined
   * 
   * @todo implement role checks
   * @todo validate token on every update (receive props) properly (against server?)
   */
  class AuthRouteWrapper extends React.Component {
    constructor(props){
      super(props)

      this.renderAuth = this.renderAuth.bind(this)
      this.renderGuest = this.renderGuest.bind(this)
      this.simpleValidate = this.simpleValidate.bind(this)

      this.state = {
        isGuest: true
      }
    }

    componentWillMount(){
      console.log("AuthRoute - Component will mount")
      // At init, we don't have the token from the store yet.
      // check local storage initially.
      const localToken = localStorage.getItem('token')
      this.simpleValidate(localToken)
    }

    componentWillReceiveProps(nextProps){
      // get token from store on subsequent updates
      const { token, dispatch } = nextProps
      this.simpleValidate(token)
    }

    /**
     * Simple client-side validation. This is NOT a secure check;
     * only for UI responsiveness. The server API handles the
     * rejection and will not deliver data without a valid
     * token either way.
     * @param {jwt token} token 
     */
    simpleValidate(token){
      const {dispatch} = this.props

      if(token){
        const decodedToken = jwtDecode(token)
        const dateNow = moment().unix()

        if(decodedToken.exp < dateNow){
          console.log("AuthRoute Update: Token expired. Is now Guest.")
          this.setState({ isGuest: true })

          dispatch({type: types.AUTH_LOGOUT_USER })
          return
        }

        console.log("AuthRoute Update: is authenticated with token")
        this.setState({ isGuest: false })

      } else {
        console.log("AuthRoute Update: is guest.")
        this.setState({ isGuest: true })
      }
    }

    renderGuest(){
      const {component, ...rest} = this.props
      return (
        <Route {...rest} render={props => (
            <Redirect to={{
              pathname: constants.LOGIN_PATH,
              state: { from: props.location }
            }}/>
        )}/>
      )
    }

    renderAuth(){
      const {component, ...rest} = this.props
      const Component = component

      return(
        <Route {...rest} render={props => (
          <Component {...props} />
      )}/>
      )
    }

    render() {
      //alert(this.state.isGuest)
      return this.state.isGuest ? this.renderGuest() : this.renderAuth()
    }
  }

  export const AuthRoute = connect( state => ({
    token: state.auth.token
  }))(AuthRouteWrapper)
  