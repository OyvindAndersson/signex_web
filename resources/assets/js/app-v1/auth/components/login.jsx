import React from 'react'

export default class Login extends React.Component {
    handleSubmit(e){
        alert(e);
    }
    render() {
        return(
            <div>
                <input type="text" name="email" placeholder="example@example.com" />
                <input type="password" name="password" />
                <button type="submit" onClick={this.handleSubmit.bind(this)}>Login</button>
            </div>
        );
    }
}