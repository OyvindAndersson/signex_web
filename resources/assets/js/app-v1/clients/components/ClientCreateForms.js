/**
 * All types of create forms for the Client model
* Simpler to organize all in one file
* @todo If components become too large, move to seperate
*/

import React from 'react'
import {connect} from 'react-redux'

import Brreg from 'AppUtils/brreg'
import { createClientAction } from 'Clients/actions'
import { isCreateClientActive, didCreateClientFail  } from 'Clients/selectors'
import { toast } from 'react-toastify'

class ClientDataSelectItem extends React.Component {
    constructor(props){
        super(props)

        this.handleDataItemSelected = this.handleDataItemSelected.bind(this)
    }

    handleDataItemSelected(e){
        if(this.props.handleDataItemSelected){
            this.props.handleDataItemSelected(this.props.itemData)
        }
    }
    render(){
        const item = this.props.itemData
        const employees = item.antallAnsatte ? item.antallAnsatte : null

        let address = null
        if(item.forretningsadresse){
            address = {
                street: item.forretningsadresse.adresse ? item.forretningsadresse.adresse : '',
                area: item.forretningsadresse.poststed ? item.forretningsadresse.poststed : ''
            }
        }

        return(
            <a className="list-group-item list-group-item-action flex-column align-items-start"
                onClick={this.handleDataItemSelected}>

                <div className="d-flex w-100 justify-content-between">
                    <small className="mb-1"><strong>{item.navn}</strong></small>
                    <small>{item.organisasjonsnummer}</small>
                </div>
                <ul className="item-data">
                    { employees ? (<li>{employees} employees</li>) : null }
                    { address ? (<li>{address.street}, {address.area}</li>) : null }
                </ul>
            </a>
        )
    }
}

/**-------------------------------------------------
 * Client Create Inline Form View
 * -------------------------------------------------
 * Inline/navbar form for quickly creating new clients
 * 
 * @todo Auto-focus / tab to list when results show, so we can navigate the results with up/down/tab keys
 */
class ClientCreateInlineFormView extends React.Component {
    constructor(props){
        super(props)

        this.handleInputChanged = this.handleInputChanged.bind(this)
        this.handleBrregCallback = this.handleBrregCallback.bind(this)
        this.handleSelectBoxItemSelected = this.handleSelectBoxItemSelected.bind(this)
        this.handleClientCreateForm = this.handleClientCreateForm.bind(this)
        this.onSuccess = this.onSuccess.bind(this)

        // handleBrregCallback called when Brreg has got a response from server
        this.brreg = new Brreg(this.handleBrregCallback)
        this.state = {
            clientName: '',
            clientOrgNr: '',
            brregResults: [],
            isSavingClient: false,
            saveClientFailed: false,
            saveSuccess: false
        }
    }

    componentWillReceiveProps(props){
        if(this.state.isSavingClient){
            this.setState({
                isSavingClient: false,
                saveSuccess: props.saveSuccess
            })

            if(props.saveSuccess){
                this.onSuccess()
            } else {
                toast.error(`Client failed to save!`)
            }
            
        }
    }

    onSuccess(){
        if(this.props.onSuccess){
            this.props.onSuccess()
        }
    }

    handleInputChanged(e){
        const {value, name} = e.target

        this.setState({
            [name]: value
        })

        // Update results. API only responds to queries with filter > 2 chars.
        if(value.length > 2){
            this.brreg.searchByName(value, 5)
        } else {
            // hack for now. Hides the result list UI
            this.handleBrregCallback(null)
        }
    
    }

    handleBrregCallback(results){
        this.setState({
            brregResults: results ? results : []
        })
    }

    handleSelectBoxItemSelected(dataItem){
        this.setState({ 
            clientName: dataItem.navn,
            clientOrgNr: dataItem.organisasjonsnummer,
            brregResults: []
        })
        // The brreg data was selected and filled in the form
        // now, clear all results and focus the submit button
        // for quick submit
        this.refs.submitBtn.focus()
    }

    handleClientCreateForm(e){
        e.preventDefault()

        this.props.createClient({
            name: this.state.clientName,
            org_nr: this.state.clientOrgNr
        })

        this.setState({
            clientName: '',
            clientOrgNr: '',
            isSavingClient: true,
            saveSuccess: false  // So we can respond in willReceiveProps to this actually happening
        })

        toast.info(`Saving...`)
    }

    render(){
        const brregData = this.state.brregResults.map((item) => {
            return (
                <ClientDataSelectItem key={item.organisasjonsnummer} 
                itemData={item} 
                handleDataItemSelected={this.handleSelectBoxItemSelected} />
            )
        })

        return(
        <form onSubmit={this.handleClientCreateForm}  className="form-inline my-2 my-lg-0">
            <div className="form-group">
                <span className="navbar-text mr-sm-2">Ny kunde i farta:</span>
                <div className="data-select">
                    <input className="form-control mr-sm-2"
                        name="clientName"
                        placeholder="Client..."
                        autoComplete="off"
                        value={this.state.clientName}
                        onChange={this.handleInputChanged} />

                    <div className="result-box" >
                        <div className="list-group">
                            {brregData}
                        </div>
                    </div>
                </div>

                <input className="form-control mr-sm-2"
                    name="clientOrgNr" 
                    placeholder="org. nr..."
                    autoComplete="off"
                    value={this.state.clientOrgNr}
                    onChange={this.handleInputChanged} />

                <button ref="submitBtn" 
                    type="submit"
                    className="btn btn-outline-success my-2 my-sm-0">Save
                </button>
            </div>
        </form>
        )
    }
}

function mapStateToProps(state){
    return {
        isSavingClient: isCreateClientActive(state),
        saveClientFailed: didCreateClientFail(state),
        saveSuccess: state.ui.clients.clientCreatedSuccess
    }
}

function mapDispatchToProps(dispatch){
    return {
        createClient: payload => dispatch(createClientAction(payload))
    }
}
 
export const ClientCreateInlineForm = connect(mapStateToProps, mapDispatchToProps)(ClientCreateInlineFormView)