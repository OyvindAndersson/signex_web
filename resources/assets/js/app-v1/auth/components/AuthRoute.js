import React from 'react'
import { Route, Redirect, withRouter } from 'react-router-dom'
import { connect } from 'react-redux'

import constants from 'Auth/constants'
import types from 'Auth/actionTypes'
import { logoutAction } from 'Auth/actions'
import { isVerifyTokenActive, verifyTokenError, isClientAuthenticated } from 'Auth/selectors'
import { onlyAuth } from './authHocs'

  /**
   * Auth Route Wrapper
   * 
   * Wraps a react-router Route. Yields access based upon
   * authenticated status and role, if defined
   * 
   */
class AuthRoute2 extends React.Component {
  constructor(props){
    super(props)

    this.renderAuth = this.renderAuth.bind(this)
    this.renderGuest = this.renderGuest.bind(this)

    this.state = {
      isAuthenticated: props.isAuthenticated ? props.isAuthenticated : false,
      hasLoadedStore: false
    }
  }

  componentWillReceiveProps(nextProps){
    //console.log("AuthRoute - Component will update")
    this.setState({
      isAuthenticated: nextProps.isAuthenticated,
      hasLoadedStore: true
    })
  }

  renderGuest(){
    const {component, ...rest} = this.props
    return (
      <Route {...rest} render={props => (
          <Redirect to={{
            pathname: constants.LOGIN_PATH,
            state: { from: props.location }
          }}/>
      )}/>
    )
  }

  renderAuth(){
    const {component, ...rest} = this.props
    const Component = component

    return(
      <Route {...rest} render={props => (
        <Component {...props} />
    )}/>
    )
  }

  render() {
    const {isAuthenticated} = this.props
    if(!this.state.hasLoadedStore && !isAuthenticated){
      return null
    }
    console.log("authRoute - isAuthenticated", `[${this.state.isAuthenticated}]`)

    return this.state.isAuthenticated ? this.renderAuth() : this.renderGuest()
  }
}

/**
 * AuthRoute
 * 
 * Wraps a react-router Route, and should be rendered
 * when - and only when - authenticated
 * 
 * @todo User role check
 */
class AuthRoute extends React.Component {
  constructor(props){
    super(props)

    var isAuthed = props.isAuthenticated || localStorage.getItem('isAuthenticated') == 'true'
    this.state = {
      isAuthenticated: isAuthed
    }
  }

  componentWillReceiveProps(nextProps){
    //console.log("AuthRoute - Component will update")
    this.setState({
      isAuthenticated: nextProps.isAuthenticated
    })
  }

  render(){
    const {component, ...rest} = this.props
    const Component = component

    return(
      <Route {...rest} render={props => (
        <Component {...props} />
    )}/>
    )
  }
}

/**
 * Component to render if not AuthRoute
 * @param {*} props 
 */
const NotAuthRoute = (props) => {
  const {component, ...rest} = props
    return (
      <Route {...rest} render={props => (
          <Redirect to={{
            pathname: constants.LOGIN_PATH,
            state: { from: props.location }
          }}/>
      )}/>
    )
}
  

function mapStateToProps(state) {
  return {
    isAuthenticated: isClientAuthenticated(state),
    isVerifyingToken: isVerifyTokenActive(state)
  }
}

function mapDispatchToProps(dispatch) {
  return ({
    logout: () => dispatch(logoutAction())
  })
}

/**
 * AuthRoute
 * 
 * Conditional rendering logic really happens in 'onlyAuth' HOC.
 * Renders wrapped component when authenticated, ElseComponent 
 * if not.
 */
export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(
    onlyAuth({ ElseComponent: NotAuthRoute })(AuthRoute)
  )
)