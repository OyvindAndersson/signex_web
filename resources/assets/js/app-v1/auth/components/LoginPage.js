import React from 'react'
import {connect} from 'react-redux'
import {Redirect, withRouter} from 'react-router-dom'

import constants from 'Auth/constants'
import { verifyLoginAction } from "Auth/actions"
import { isVerifyLoginActive, isClientAuthenticated, isVerifyTokenActive } from "Auth/selectors";
import {exceptAuth} from './authHocs'
import LoginForm from './LoginForm'



/**
 * Container component for the login page
 */
class LoginPage extends React.Component {
    constructor(props){
        super(props)
        this.state = { isVerifyingLogin: false }
    }
    componentWillReceiveProps(nextProps){
        this.setState({ isVerifyingLogin: nextProps.isVerifyingLogin })
    }
    render(){
        return(<LoginForm authLoginUser={this.props.verifyLogin} isVerifyingLogin={this.state.isVerifyingLogin} /> )
    }
}

/**
 * The component to render if/when authentication is successful.
 * @param {*} props 
 */
const LoginSuccessful = (props) => {
    const {from} = props.location.state || { from: { pathname: constants.AUTH_ROOT_PATH }}

    if(from.pathname && from.pathname.indexOf('logout') !== -1){
        return <Redirect to={constants.AUTH_ROOT_PATH} />
    } else {
        return <Redirect to={from} />
    }
}


function mapStateToProps(state) {
    return {
        isAuthenticated: isClientAuthenticated(state),
        isVerifyingLogin: isVerifyLoginActive(state),
        isVerifyingToken: isVerifyTokenActive(state)
    }
}

function mapDispatchToProps(dispatch) {
    return {
        verifyLogin: credentials => dispatch(verifyLoginAction( credentials ))
    }
}

export default withRouter(
    connect(mapStateToProps, mapDispatchToProps)(
        exceptAuth({ ElseComponent: LoginSuccessful })(LoginPage)
    )
)