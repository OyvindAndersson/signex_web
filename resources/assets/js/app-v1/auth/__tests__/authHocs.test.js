import React from 'react'
import { connect } from "react-redux"
import configureMockStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'

import { onlyAuth, exceptAuth } from '../components/authHocs'

describe('onlyAuth HOC', () => {
    let MainComponent
    let LoadingComponent
    let ElseComponent
    let WrappedComponent

    beforeEach(() => {
        MainComponent = () => (<div>MainComponent</div>)
        LoadingComponent = () => (<div>LoadingComponent</div>)
        ElseComponent = () => (<div>ElseComponent</div>)
        
        WrappedComponent = connect((state) => ({
            isAuthenticated: state.auth.isAuthenticated,
            isVerifyingLogin: state.auth.isVerifyingLogin,
            isVerifyingToken: state.auth.isVerifyingToken
        }))(onlyAuth({ LoadingComponent, ElseComponent })(MainComponent))
    })

    it('should render MainComponent when only isAuthenticated is true and isVerifyingLogin and isVerifyingToken are both false', () => {
        
        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: true, 
                isVerifyingLogin: false,
                isVerifyingToken: false
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })

    it('should render LoadingComponent when both isVerifyingLogin and isAuthenticated are true and only isVerifyingToken is false', () => {

        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: true, 
                isVerifyingLogin: true,
                isVerifyingToken: false
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })

    it('should render LoadingComponent when both isVerifyingToken and isAuthenticated are true and only isVerifyingLogin is false', () => {

        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: true, 
                isVerifyingLogin: false,
                isVerifyingToken: true
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })

    it('should render LoadingComponent when only isVerifyingToken is true and isAuthenticated and isVerifyingLogin are both false', () => {

        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: false, 
                isVerifyingLogin: false,
                isVerifyingToken: true
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })

    it('should render LoadingComponent when only isVerifyingLogin is true and isAuthenticated and isVerifyingToken are both false', () => {

        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: false, 
                isVerifyingLogin: true,
                isVerifyingToken: false
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })

    it('should render LoadingComponent when isVerifyingLogin and isAuthenticated and isVerifyingToken are all true', () => {

        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: true, 
                isVerifyingLogin: true,
                isVerifyingToken: true
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })

    it('should render ElseComponent when isAuthenticated and isVerifyingLogin and isVerifyingToken are all false', () => {

        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: false, 
                isVerifyingLogin: false,
                isVerifyingToken: false
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })
})

describe('exceptAuth HOC', () => {
    let MainComponent
    let LoadingComponent
    let ElseComponent
    let WrappedComponent

    beforeEach(() => {
        MainComponent = () => (<div>MainComponent</div>)
        LoadingComponent = () => (<div>LoadingComponent</div>)
        ElseComponent = () => (<div>ElseComponent</div>)
        
        WrappedComponent = connect((state) => ({
            isAuthenticated: state.auth.isAuthenticated,
            isVerifyingLogin: state.auth.isVerifyingLogin,
            isVerifyingToken: state.auth.isVerifyingToken
        }))(exceptAuth({ LoadingComponent, ElseComponent })(MainComponent))
    })

    it('should render MainComponent when isAuthenticated and isVerifyingLogin and isVerifyingToken are all false', () => {
        
        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: false, 
                isVerifyingLogin: false,
                isVerifyingToken: false
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })

    it('should render LoadingComponent when both isVerifyingLogin and isAuthenticated are true, and only isVerifyingToken is false', () => {

        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: true, 
                isVerifyingLogin: true,
                isVerifyingToken: false
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })

    it('should render LoadingComponent when both isVerifyingToken and isAuthenticated are true, and only isVerifyingLogin is false', () => {

        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: true, 
                isVerifyingLogin: false,
                isVerifyingToken: true
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })

    it('should render LoadingComponent when only isVerifyingToken is true and isAuthenticated and isVerifyingLogin are both false', () => {

        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: false, 
                isVerifyingLogin: false,
                isVerifyingToken: true
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })

    it('should render LoadingComponent when only isVerifyingLogin is true and isAuthenticated and isVerifyingToken are both false', () => {

        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: false, 
                isVerifyingLogin: true,
                isVerifyingToken: false
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })

    it('should render LoadingComponent when isVerifyingLogin and isAuthenticated and isVerifyingToken are all true', () => {

        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: true, 
                isVerifyingLogin: true,
                isVerifyingToken: true
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })

    it('should render ElseComponent when only isAuthenticated is true and isVerifyingLogin and isVerifyingToken are both false', () => {

        const mockStore = configureMockStore()({
            auth: { 
                isAuthenticated: true, 
                isVerifyingLogin: false,
                isVerifyingToken: false
            }
        });

        let tree = renderer.create(
            <Provider store={mockStore}>
                <WrappedComponent location={ {state: { from: {pathname: "test"}}}} />
            </Provider>
        ).toJSON()
        
        expect(tree).toMatchSnapshot()
    })
})