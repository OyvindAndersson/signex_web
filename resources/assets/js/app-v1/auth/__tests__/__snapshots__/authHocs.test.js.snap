// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exceptAuth HOC should render ElseComponent when only isAuthenticated is true and isVerifyingLogin and isVerifyingToken are both false 1`] = `
<div>
  ElseComponent
</div>
`;

exports[`exceptAuth HOC should render LoadingComponent when both isVerifyingLogin and isAuthenticated are true, and only isVerifyingToken is false 1`] = `
<div>
  LoadingComponent
</div>
`;

exports[`exceptAuth HOC should render LoadingComponent when both isVerifyingToken and isAuthenticated are true, and only isVerifyingLogin is false 1`] = `
<div>
  LoadingComponent
</div>
`;

exports[`exceptAuth HOC should render LoadingComponent when isVerifyingLogin and isAuthenticated and isVerifyingToken are all true 1`] = `
<div>
  LoadingComponent
</div>
`;

exports[`exceptAuth HOC should render LoadingComponent when only isVerifyingLogin is true and isAuthenticated and isVerifyingToken are both false 1`] = `
<div>
  LoadingComponent
</div>
`;

exports[`exceptAuth HOC should render LoadingComponent when only isVerifyingToken is true and isAuthenticated and isVerifyingLogin are both false 1`] = `
<div>
  LoadingComponent
</div>
`;

exports[`exceptAuth HOC should render MainComponent when isAuthenticated and isVerifyingLogin and isVerifyingToken are all false 1`] = `
<div>
  MainComponent
</div>
`;

exports[`onlyAuth HOC should render ElseComponent when isAuthenticated and isVerifyingLogin and isVerifyingToken are all false 1`] = `
<div>
  ElseComponent
</div>
`;

exports[`onlyAuth HOC should render LoadingComponent when both isVerifyingLogin and isAuthenticated are true and only isVerifyingToken is false 1`] = `
<div>
  LoadingComponent
</div>
`;

exports[`onlyAuth HOC should render LoadingComponent when both isVerifyingToken and isAuthenticated are true and only isVerifyingLogin is false 1`] = `
<div>
  LoadingComponent
</div>
`;

exports[`onlyAuth HOC should render LoadingComponent when isVerifyingLogin and isAuthenticated and isVerifyingToken are all true 1`] = `
<div>
  LoadingComponent
</div>
`;

exports[`onlyAuth HOC should render LoadingComponent when only isVerifyingLogin is true and isAuthenticated and isVerifyingToken are both false 1`] = `
<div>
  LoadingComponent
</div>
`;

exports[`onlyAuth HOC should render LoadingComponent when only isVerifyingToken is true and isAuthenticated and isVerifyingLogin are both false 1`] = `
<div>
  LoadingComponent
</div>
`;

exports[`onlyAuth HOC should render MainComponent when only isAuthenticated is true and isVerifyingLogin and isVerifyingToken are both false 1`] = `
<div>
  MainComponent
</div>
`;
